Phase 1 Pixel DQM

schneiml opened the pull request:
This is a test/discussion PR. I am not sure whether it can/should/will be merged. @fioriNTU FYI.

This introduces the shared code for the new, geometry-independent Pixel DQM (atm called Phase1, though it should be very capable of also handling the old detector. It might even handle Strips, but this is all out of scope for now). This is still very rough and not fully implemented, but the supplied sample analyzer already produces some plots on digis.

This code relies on Max Stark's ( @DirtyDan88 ) work in the Alignment package. The commits are included in this PR, but they should probably be in a separate one. However, I could not find any. @mmusich @tlampen are there any plans on merging this stuff?

To make matters worse, I had to apply some changes to the Alignment package in commit https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e . These are untested, and might break things.

On more general things:
-  I don't know yet how the new could could and should be integrated in tests. Atm there are no automated tests.
- The code liberally uses `assert` to ensure it is used correctly. These should all be situations that only appear due to coding errors, but I still wonder whether exceptions are preferred in this case (backtraces might be useful).
- The shared code does not introduce a namespace. I think it could be a good idea, but I could not deduce the typical rules for when to use one (and how to call it).

DirtyDan88 committed:
added structure-names for upgraded geometry

DirtyDan88 committed:
implemented AlignmentLevels, which defines the granularity of CompositeAlignables

DirtyDan88 committed:
added structure-names for TIB and TEC in PhaseII (marked as invalid)

DirtyDan88 committed:
removed some hardcoded geometry assumptions

DirtyDan88 committed:
added top-level structures to TrackerCounters

DirtyDan88 committed:
for simplicity: renamed AlignSetup -> AlignableMap and removed template param (was only used for one class)

DirtyDan88 committed:
renamed indexing-system (class Counters -> AlignableIndexer)

DirtyDan88 committed:
new, geometry-independent AlignableComposite-builder; builds tracker-Alignables depending on the currently loaded geometry

DirtyDan88 committed:
implemented test analyzer for geometry-analysis and alignable-creation

schneiml committed:
Merged refs/pull/13590/head from repository cms-sw

schneiml committed:
A from-scratch DQM Module

schneiml committed:
Minimal DQM config

schneiml committed:
Customization to turn the new DQM on

schneiml committed:
Switch to sim digis; a thing to b centralized...

schneiml committed:
A bit of tracing output

schneiml committed:
PixelDQM framework skeleton. The hard stuff is not implemented yet, but the skeleton compiles.

schneiml committed:
FIx buildfile

schneiml committed:
TopFolder option for manager

schneiml committed:
Use Manager for Digis

schneiml committed:
Specify the SummationSpecification more precisely

schneiml committed:
A test harness that does not do the full DQM sequence

schneiml committed:
Simple booking. Not optimal yet, hardcoded ranges etc... but generates a plot\!

schneiml committed:
The actual test harness (file was missing in prev. commit)

schneiml committed:
WIP: read structure from alignment

schneiml committed:
Merged AlignmentGeometryUpgrade_800pre6 from repository DirtyDan88

This should be merged indepedently. Turns this branch into a bit of an Frankenbranch now.

schneiml committed:
Fix compilation problems with the new Alignment code.

Mostly minor, but the re-added trackerTopology() accessor might return an invalid pointer, idk.

schneiml committed:
WIP: Bitmap Extraction. Results for Pixel are still implausible

schneiml committed:
Some hierarchies of histograms. The Geometry seems implausible, needs to be checked.

schneiml committed:
Slightly modfied bitmasks.

Now the Geometry looks more sane but it does not resolve the Half* structures anymore; however I can't see manually how a bitmask could tell this.

schneiml committed:
Enable config for step2

schneiml committed:
Split into step1/step2 using a template skeleton.

This is less elegant than I hoped, but in return it gives us the power to remove most duplicated code. Note the ugly workaround needed for consumes<...>.

schneiml committed:
Allow for multiple histograms in one (Base-)Plugin

schneiml committed:
Make the interfaces a bit nicer

schneiml committed:
Implement more spec steps, and move the Builder impl out of the header.

schneiml committed:
Move module list to GeometryInterface, fix label problem

schneiml committed:
First working step2 histograms

schneiml committed:
Make ranges setable

schneiml committed:
Working EXTEND_X summation.

schneiml committed:
Improve metadata tracking. Now everything is a TH1D.

schneiml committed:
Switch to proper logging

schneiml committed:
Merged phase1pixeldqm-framework-intrusive from repository schneiml

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for CMSSW_8_1_X.

It involves the following packages:

Alignment/CommonAlignment
Alignment/CommonAlignmentAlgorithm
Alignment/LaserAlignment
Alignment/TrackerAlignment
DQM/SiPixelPhase1Common
DQM/SiPixelPhase1Digis
SLHCUpgradeSimulations/Configuration


The following packages do not have a category, yet:

DQM/SiPixelPhase1Common
DQM/SiPixelPhase1Digis

@civanch, @diguida, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @mmusich, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @pakhotin, @makortel, @tocheng, @Martin-Grunewald, @tlampen, @mschrode, @mmusich, @threus this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

mmusich commented:
@schneiml we definitely plan to merge Max stuff (needed for next year alignment), but why can't that go separately? If you have dependencies can we build them up in a staggered approach?

mmusich commented:
@smuzaffar the new packages should have DQM signature
@vanbesien @deguio

makortel commented on SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py:
```diff
@@ -143,7 +143,9 @@ def customise_DQM(process,pileup):
-    process.DQMOfflinePreDPG.remove(process.siPixelOfflineDQM_source)
+    old = process.siPixelOfflineDQM_source
+    process.load('DQM.SiPixelPhase1Common.SiPixelPhase1OfflineDQM_source_cff')
+    process.DQMOfflinePreDPG.replace(old, process.siPixelPhase1OfflineDQM_source)
```
@schneiml At this point please don't add new customizations to `phase1TkCustoms` (it's being cleaned up). Use the era mechanism instead, and delete the line removing `siPixelOfflineDQM_source` from the `DQMOfflinePreDPG` sequence.

Essentially you probably want to use `eras.phase1Pixel.toReplaceWith()` in https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DQM/SiPixelCommon/python/SiPixelOfflineDQM_source_cff.py to replace the contents of `siPixelOfflineDQMSource`. See e.g. https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoTracker/IterativeTracking/python/iterativeTk_cff.py or https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py for practical examples (in tracking configuration).

Said that, for me these changes are fine for testing or RFC-PRs (like this one seems to be at the moment), but for integration an era-based solution is needed.

makortel commented on SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py:
```diff
@@ -259,6 +261,16 @@ def customise_harvesting(process):
+    # Pixel DQM needs to be updated for phase1
+    #old = process.PixelOfflineDQMClientNoDataCertification
+    process.load('DQM.SiPixelPhase1Common.SiPixelPhase1OfflineDQM_harvesting_cff')
+    process.PixelOfflineDQMClientNoDataCertification.insert(1, process.siPixelPhase1OfflineDQM_harvesting)
```
@schneiml Also this customization should eventually be done with eras.

schneiml commented:
@mmusich I think we don't understand each other correctly. I would like to have the alignment stuff in a dedicated PR, but I don't feel I am the right person to open one. Esp. regarding the changes in https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e . 

And my code now depends on Max's code, so I cannot open a PR w/o these commits somewhere. I guess the right approach is doing a new PR for the Alignment code, get it merged, and then I rebase/reopen this PR on top.

@makortel Thanks for the hints. The entire configuration stuff in this PR is suboptimal, and I don't understand enough yet to fix it. I'll look into it.

mmusich commented:
@schneiml thanks for the clarification. 
Can @DirtyDan88 or any Tracker alignment group representative open a PR with his commits + https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e  ?

mmusich commented:
@mschrode @ghellwig @tlampen see above comment

smuzaffar commented:
@schneiml, there is cyclic dependency b/w DQM/SiPixelPhase1Common & DQM/SiPixelPhase1Digis
(BuildFile and python cfi files). Do you really want to have SiPixelPhase1Digis with one plugin? I would suggest that move it in DQM/SiPixelPhase1Common/plugins directory

schneiml commented:
@smuzaffar  the same sort of dependency exists in the old `DQM/SiPixelCommon`. In the end, there should be more `SiPixelPhase1*` packages like the digi one, most likely also some that have more than one plugin. 

This sounds reasonable to me, but the current SiPixel DQM is not necessarily known for good design. Would you recommend having the full Pixel DQM in one package? It should be much less code than the old one. I like the idea, it could reduce the amount of boilerplate to add a new plugin...

schneiml committed:
Configure the message logger.

The warning level is spammed by cabling, but supressing seems to cause a crash. Also LogDebug behaves differently from the higher ones, so I stick to Info for some debug output.

smuzaffar commented:
If there are more SiPixelPhase1* like packages to come then I would suggest to design the packages structure now.

I do not mind having one DQM Pixal package as long as their external dependency is in control e.g. plugin with same set of dependency could live together.

schneiml committed:
Implement step1 step execution for the basic cases, with per-bx counts as an example. Lots of the new stuff are not well tested yet.

schneiml committed:
Implement string interening and or-columns.

Interning can be considered a premature optimization, but throwing around std::string in the hottest places is more of a design problem...

schneiml committed:
Use or-columns

schneiml committed:
Implement more 2D stuff. step2 EXTEND_Y is still missing.

schneiml committed:
Add columns for row/col. We need a way to have geometry-dependent parameters there.

schneiml committed:
Use the new 2D features for some hitmaps.

schneiml committed:
Merged phase1pixeldqm-framework-intrusive from repository schneiml

cmsbuild commented:
Pull request #13798 was updated. @civanch, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @mmusich, @davidlange6 can you please check and sign again.

mmusich commented:
@ghellwig I think we agreed a new PR with the alignment only commits should have appeared early this week. is there any showstopper?

ghellwig commented:
We are currently running a check to test Multi-IOV and full module-level alignment.
While doing that we fixed another bug in the new alignment framework.
We hope to get the comparison of old and new framework for the current geometry done within this week.
Btw. we had also fixed the same thing as the commit https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e, which @schneiml referenced.

ghellwig commented:
@mmusich, @schneiml 
We have created a PR (#13976) with Max changes, also including your fix in https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e

Note, that it has been rebased to the tip of official-cmssw/CMSSW_8_0_X, so the commit IDs have changed.

mmusich commented:
@ghellwig thanks, let's get the alignment stuff in first then.

mmusich commented:
@schneiml 
https://github.com/cms-sw/cmssw/pull/13976 has been merged in the meanwhile. 
Do we still need this one? The branch doesn't merge anymore

schneiml commented:
@mmusich no, this PR can be closed, #14230 is the current one.

Also, the code interfacing to Alignment is still present but `#if 0`'ed in the new PR since we now use our own naming convention. It might still be useful for testing now and then, but in the normal case the new PR is independent of alignment.
